timeline/src/Data/Timeline/Internal.hs

```diff
@@ -0,0 +1,299 @@
+  -}
+  { initialValue :: a,
+    -- | changes are keyed by their "effective from" time, for easier lookup
+    values :: Map UTCTime a
+  }
+  deriving stock (Eq, Generic, Functor, Foldable, Traversable)
+
+fromValues ::
+  -- | initial value
+  a ->
+  -- | new values begin to take effect at the specified times
+  Map UTCTime a ->
+  Timeline a
+fromValues initialValue values = Timeline {initialValue, values}
+
+instance Applicative Timeline where
+  pure :: a -> Timeline a
+  pure a = Timeline {initialValue = a, values = mempty}
+
+  (<*>) :: forall a b. Timeline (a -> b) -> Timeline a -> Timeline b
+  fs@Timeline {initialValue = initialFunc, values = funcs} <*> xs@Timeline {initialValue, values} =
+    Timeline
+      { initialValue = initialFunc initialValue,
+        values = mergedValues
+      }
+    where
+      mergedValues :: Map UTCTime b
+      mergedValues =
+        Map.merge
+          (Map.mapMissing $ \t f -> f $ peek xs t)
+          (Map.mapMissing $ \t x -> peek fs t x)
+          (Map.zipWithMatched (const ($)))
+          funcs
+          values
+
+instance (Show a) => Show (Timeline a) where
+  show Timeline {initialValue, values} =
+    unlines $
+      "\n----------Timeline--Start-------------"
+        : ("initial value:                 " <> show initialValue)
+        : fmap showOneChange (Map.toAscList values)
+        ++ ["----------Timeline--End---------------"]
+    where
+      showOneChange :: (UTCTime, a) -> String
+      showOneChange (t, x) = "since " <> show t <> ": " <> show x
+
+peek :: Timeline a -> UTCTime -> a
+peek Timeline {..} time = maybe initialValue snd $ Map.lookupLE time values
+
+data TimeRange = TimeRange
+  { trFrom :: Maybe UTCTime,
+    trTo :: Maybe UTCTime
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+
+isTimeAfterRange :: UTCTime -> TimeRange -> Bool
+isTimeAfterRange t TimeRange {trTo} = maybe False (t >=) trTo
+
+instance FunctorWithIndex TimeRange Timeline where
+  imap :: (TimeRange -> a -> b) -> Timeline a -> Timeline b
+  imap f Timeline {..} =
+    Timeline
+      { initialValue = f initialRange initialValue,
+        values = flip Map.mapWithKey values $ \from value ->
+          let timeRange = TimeRange (Just from) (fst <$> Map.lookupGT from values)
+           in f timeRange value
+      }
+    where
+      initialRange = TimeRange Nothing $ fst <$> Map.lookupMin values
+
+instance FoldableWithIndex TimeRange Timeline
+
+instance TraversableWithIndex TimeRange Timeline where
+  itraverse :: (Applicative f) => (TimeRange -> a -> f b) -> Timeline a -> f (Timeline b)
+  itraverse f = sequenceA . imap f
+
+changes :: Timeline a -> Set UTCTime
+changes Timeline {values} = Map.keysSet values
+
+data Record a = Record
+  { effectiveFrom :: UTCTime,
+    effectiveTo :: Maybe UTCTime,
+    value :: a
+  }
+  deriving stock (Eq, Functor)
+
+type Getter s a = forall f. (Contravariant f) => (a -> f a) -> s -> f s
+
+recordEffectiveFrom :: Getter (Record a) UTCTime
```

Comment:
Why are we providing handwritten optics? I'm in favour of the `generic-lens` approach, in which case the handwritten optics are redundant. If the generic lenses are not available or lawful, we should document that.

Reply:
I want to keep the fields of `Record` readonly, and only allow constructing one from the smart constructor. Do you have better solution to this?

Reply:
Export the field selectors but not the constructor, and require callers to use their optics package's version of `to`?

Reply:
(As-is, `optics` users will have to convert from a van Laarhoven `Getter`.)

Reply:
Record update syntax still works even though only the type name and field selector functions are exported. (This is actually described in [yesod book](https://www.yesodweb.com/book/settings-types))

So I exported `getXXX` functions instead of field selector functions.

Reply:
Just when I think Haskell records couldn't get weirder...